text/2019-11-13-follower-replication.md

```diff
@@ -0,0 +1,104 @@
+# Follower Replication
+
+## Summary
+
+This RFC introduces a new mechanism for Raft protocol which allows a follower to send raft logs and snapshots to other followers. The target of this feature is to reduce network transmission cost between different data centers in Log Replication.
+
+## Motivation
+
+In the origin Raft protocol, a follower can only receive raft logs and snapshots from its leader, which could be insufficient in some situations. For example, when a Raft group is distributed in different data centers, log replication between a new node and the leader could be expensive if they are in different data centers. In this case, internal follower-to-follower transfer in one data center can be far more efficient than the traditionally stipulated leader-to-follower transfer. And in a cluster with massive nodes, the leader can have lower system load due to the less operations in network transmission.
+
+## Detailed design
+
+There are four key concepts of design:
+
+- Every peer in a Raft group is associated with a `group_id`.
```

Comment:
I was actually thinking whether we could make the delegation group support native in the first version. Suppose we want to form another delegation group in runtime, we need to change the static configs and do the cluster rolling bounce. 

While it makes sense for the first version to focus on static configs, we may also propose runtime formulation of delegation group through third-party control such as PD. When the leader load goes up, it makes sense to distribute its load to other up-to-date followers. We could briefly talk about how we could keep the door open for such a future improvement, with the first version only supporting static `group_id`.

Reply:
Good idea! Actually it's the first design of how the leader manages all the group info.